ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs

```diff
@@ -200,32 +209,29 @@ initialTestLedgerState = UTxTok {
 -- | Get the token value associated to a given token. This is coupled to the
 -- generators semantics.
 pointTValue :: Token -> TValue
-pointTValue (Token GenesisPoint)        = TValue 0
-pointTValue (Token (BlockPoint slot _)) = TValue slot
+pointTValue = TValue . pointSlot . unToken
 
 genBlocks ::
-     ExtLedgerCfg TestBlock
-  -> Word64
+     Word64
   -> Point TestBlock
   -> [TestBlock]
-genBlocks _   0 _ = []
-genBlocks cfg n pt =
-  let b  = genBlock pt
-      bs = genBlocks cfg (n - 1) (blockPoint b)
-  in b:bs
+genBlocks n pt0 = take (fromIntegral n) (go pt0)
+  where
+    go pt = let b = genBlock pt in b : go (blockPoint b)
 
 genBlock ::
      Point TestBlock -> TestBlock
-genBlock pt = completeBlock (mkBlockFrom pt)
+genBlock pt =
+  let blk = mkBlockFrom pt
```

Comment:
The circular definition is a little concerning. I like it as a spec, but one stray `!` added somewhere and we'll have nonproductive indefinite recursion.

Possible alternatives:

  * how involved is the block -> point logic? Maybe factor it out so we can call it even before we have the block?
  * create a dummy block with no payload, calculate its point, and then cast the point

Reply:
Although I see your concern I find it a bit sad that we can't rely on laziness to solve this problem. That said I should be able to come up with a proposal for implementing the first alternative, namely: 

> how involved is the block -> point logic? Maybe factor it out so we can call it even before we have the block?

Reply:
Actually, I don't see how I could factor out the block and point logic without exposing the `TestBlock` constructors (which were not exposed in `master` and I think for a good reason). My first attempt was to write the following:

```haskell
firstBlockWithPayload :: Word64 -> ptype -> TestBlockWith ptype
firstBlockWithPayload forkNo payload = TestBlockWith
    { tbHash    = firstHash forkNo
    , tbSlot    = firstSlot
    , tbValid   = True
    , tbPayload = payload
    }

-- | Create the first hash in the given fork.
firstHash :: Word64 -> TestHash
firstHash forkNo = TestHash (forkNo NE.:| [])

firstSlot :: SlotNo
firstSlot = 1
```

I find this to be a bad solution. I could use `firstHash` and `firstSlot` in `genBlock`, however:

- I need to export the `TestBlock` constructor.
- I'm duplicating the logic of `firstBlockWithPayload` 

That's why the use of laziness was so appealing: it allows me to preserve the abstraction of how slots and hashes are determined. I'll try the second alternative, which I like the least, but should work.

Do you think a stray `!` is a big risk? I was trying to introduce an infinite loop without success (by adding the bang pattern in several places like `blk` and `tbPayload`). If we could identify the places where the introduction of `!` would cause a problem, wouldn't a comment with a warning be enough to prevent a developer from wasting her time?

Reply:
[Here](https://github.com/input-output-hk/ouroboros-network/pull/3610/commits/878848e336a588b2576a94987a6bd3b7849c2b87) is the non-circular definition.

Reply:
I was thinking
I think we can avoid the `undefined`s with:

```
 pt' = castPoint (blockPoint blk) :: Point (TestBlockWithPayload ())
 blk = mkBlockFrom pt () :: TestBlock
```

Reply:
New commit looks good; Resolving this Conversation.